File: e2e-verify.py
"""E2E 검증: Sonnet-only 풍부한 운세 결과 Playwright 테스트"""
import re
import sys
from playwright.sync_api import sync_playwright

BASE_URL = "https://fortunova.molidae.site"


def warm_cache():
    """curl로 서버 캐시를 미리 채워 504 Gateway Timeout 방지"""
    import subprocess
    print("[0/6] 서버 캐시 워밍업 (최대 150초)...")
    try:
        # 리버스 프록시 타임아웃 회피: Docker 네트워크를 통해 직접 접근
        result = subprocess.run(
            ["docker", "exec", "fortunova", "wget", "-q", "-O", "/dev/null",
             "--post-data", "year=1988&month=7&day=22&hour=10&gender=M&calendarType=solar&category=daily",
             "http://localhost:3000/partials/fortune-result"],
            capture_output=True, text=True, timeout=160,
            env={**__import__('os').environ, "DOCKER_API_VERSION": "1.43"}
        )
        rc = result.returncode
        print(f"  → exit code: {rc}")
        if rc == 0:
            print("  ✓ 캐시 워밍업 완료")
            return True
        else:
            print(f"  ⚠ 워밍업 응답: stderr={result.stderr[:200]}")
            return False
    except Exception as e:
        print(f"  ⚠ 워밍업 실패: {e}")
        return False


def verify():
    warm_cache()
    errors = []
    with sync_playwright() as p:
        browser = p.chromium.launch(headless=True)
        page = browser.new_page()

        page.on("console", lambda msg: (errors.append(f"Console[{msg.type}]: {msg.text}") if msg.type in ("error", "warning") else None))
        page.on("pageerror", lambda err: errors.append(f"Page: {err.message}"))
        # 네트워크 요청/응답 모니터링
        page.on("requestfailed", lambda req: errors.append(f"Request failed: {req.url} - {req.failure}"))

        try:
            # 1. 메인 페이지 로드
            print("[1/6] 메인 페이지 로드...")
            resp = page.goto(BASE_URL, wait_until="networkidle")
            assert resp.ok, f"HTTP {resp.status}"
            print("  ✓ 정상 로드")

            # 2. 폼 요소 확인
            print("[2/6] 폼 요소 확인...")
            page.wait_for_selector('form[hx-post="/partials/fortune-result"]')
            print("  ✓ 폼 존재")

            # 3. 폼 입력
            print("[3/6] 폼 입력...")
            page.select_option('select[name="year"]', "1988")
            page.select_option('select[name="month"]', "7")
            page.select_option('select[name="day"]', "22")
            page.select_option('select[name="hour"]', "10")
            page.check('input[name="gender"][value="M"]')
            page.select_option('select[name="category"]', "daily")
            print("  ✓ 입력 완료")

            # 4. localStorage 캐시 클리어 + 제출
            print("[4/6] 운세보기 제출 (최대 120초 대기)...")
            page.evaluate("""() => {
                for (let i = localStorage.length - 1; i >= 0; i--) {
                    const k = localStorage.key(i);
                    if (k && k.startsWith('fortunova_')) localStorage.removeItem(k);
                }
            }""")
            page.click('button[type="submit"]')
            # 네트워크 응답 대기 + DOM 변경 대기
            try:
                page.wait_for_selector("#result > div", timeout=200000)
            except Exception as wait_err:
                # 타임아웃 시 디버그 정보 수집
                result_html = page.inner_html("#result")
                print(f"  → #result HTML 길이: {len(result_html)}")
                print(f"  → #result 내용: {result_html[:500]}")
                print(f"  → 수집된 에러: {errors}")
                # HTMX 상태 확인
                htmx_check = page.evaluate("typeof htmx !== 'undefined' ? 'loaded' : 'not loaded'")
                print(f"  → HTMX 상태: {htmx_check}")
                raise wait_err
            print("  ✓ 결과 수신")

            # 결과 HTML 저장 (디버그용)
            debug_html = page.inner_html("#result")
            with open("/tmp/fortunova-e2e-result.html", "w") as f:
                f.write(debug_html)
            print(f"  → 결과 HTML 길이: {len(debug_html)} chars")

            # 5. UX 검증
            print("[5/6] UX 검증...")
            form_collapsed = page.evaluate("document.getElementById('form-section')?.classList.contains('collapsed')")
            print(f"  {'✓' if form_collapsed else '✗'} 폼 접힘: {form_collapsed}")

            summary_visible = page.evaluate("document.getElementById('form-summary')?.style.display !== 'none'")
            print(f"  {'✓' if summary_visible else '✗'} 요약 바: {summary_visible}")

            # 6. 풍부한 결과 검증
            print("[6/6] 풍부한 운세 결과 검증...")
            result_html = page.inner_html("#result")
            result_text = page.inner_text("#result")

            if "LLM_UNAVAILABLE" in result_html or "VALIDATION_ERROR" in result_html:
                raise Exception(f"에러 응답: {result_html[:300]}")
            if "파싱에 실패" in result_html:
                # HTML 저장
                with open("/tmp/fortunova-e2e-result.html", "w") as f:
                    f.write(result_html)
                raise Exception("LLM 응답 파싱 실패")

            checks = []

            def check(name, passed):
                checks.append((name, passed))
                print(f"  {'✓' if passed else '✗'} {name}")

            # 종합 점수
            score_match = re.search(r"(\d+)\s*점", result_text)
            check("종합 점수", bool(score_match))
            if score_match:
                print(f"    → {score_match.group(1)}점")

            # 핵심 섹션들
            check("상세 설명 (오늘의 운세)", "오늘의 운세" in result_text)
            check("오행 흐름", "오행 흐름" in result_text)
            check("오늘의 조언", "오늘의 조언" in result_text)
            check("오늘의 팁", "✨" in result_html)

            # 세부운
            check("세부 운세 그리드", "세부 운세" in result_text)
            check("재물운", "재물운" in result_text)
            check("건강운", "건강운" in result_text)
            check("연애운", "연애운" in result_text)
            check("직장운", "직장운" in result_text)

            # 오행 해설
            check("오행 해설", "오행 해설" in result_text)

            # 행운 정보
            check("행운 정보", "행운" in result_text)
            check("행운의 색", "행운의 색" in result_text)
            check("행운의 숫자", "행운의 숫자" in result_text)
            check("행운의 방위", "행운의 방위" in result_text)
            check("행운의 시간", "행운의 시간" in result_text)

            # 주의사항
            check("주의사항", "주의사항" in result_text)

            # 3개월 운세 흐름
            check("3개월 운세 흐름", "3개월 운세 흐름" in result_text)

            # 궁합 팁
            check("궁합/인간관계 팁", "궁합" in result_text or "인간관계" in result_text)

            # 대운 해석
            check("대운 해석", "대운 해석" in result_text)

            # 격언
            check("격언 카드", "proverb-card" in result_html)

            # 사주 정보
            check("사주 정보", "사주:" in result_text)

            # 별점
            check("별점 표시", "★" in result_html)

            # 남은 횟수
            check("남은 무료 횟수", "무료 횟수" in result_text)

            # 결과 집계
            passed = sum(1 for _, p in checks if p)
            total = len(checks)
            print(f"\n📊 결과: {passed}/{total} 항목 통과")

            if passed < total * 0.7:
                page.screenshot(path="/tmp/fortunova-e2e-fail.png", full_page=True)
                with open("/tmp/fortunova-e2e-result.html", "w") as f:
                    f.write(result_html)
                raise Exception(f"풍부한 결과 검증 실패: {passed}/{total} (70% 미달)")

            if errors:
                print(f"\n⚠ 콘솔 에러 {len(errors)}건:")
                for e in errors:
                    print(f"  - {e}")

            print("\n✅ E2E 검증 통과!")

        except Exception as ex:
            print(f"\n❌ E2E 검증 실패: {ex}")
            page.screenshot(path="/tmp/fortunova-e2e-fail.png", full_page=True)
            print("  스크린샷: /tmp/fortunova-e2e-fail.png")
            sys.exit(1)
        finally:
            browser.close()


if __name__ == "__main__":
    verify()

File: e2e-verify.mjs
/**
 * E2E 검증: Sonnet-only 풍부한 운세 결과 Playwright 테스트
 * Generator-Critique 루프: 페이지 로드 → 폼 입력 → 제출 → 풍부한 결과 검증
 */
import { chromium } from 'playwright';

const BASE_URL = 'https://fortunova.interfn.com';

async function verify() {
  const browser = await chromium.launch({ headless: true });
  const page = await browser.newPage();
  const errors = [];

  page.on('console', msg => {
    if (msg.type() === 'error') errors.push(`Console: ${msg.text()}`);
  });
  page.on('pageerror', err => errors.push(`Page: ${err.message}`));

  try {
    // 1. 메인 페이지 로드
    console.log('[1/6] 메인 페이지 로드...');
    const response = await page.goto(BASE_URL, { waitUntil: 'networkidle' });
    if (!response.ok()) throw new Error(`HTTP ${response.status()}`);
    console.log('  ✓ 정상 로드');

    // 2. 폼 요소 확인
    console.log('[2/6] 폼 요소 확인...');
    await page.waitForSelector('form[hx-post="/partials/fortune-result"]');
    console.log('  ✓ 폼 존재');

    // 3. 폼 입력
    console.log('[3/6] 폼 입력...');
    await page.selectOption('select[name="year"]', '1988');
    await page.selectOption('select[name="month"]', '7');
    await page.selectOption('select[name="day"]', '22');
    await page.selectOption('select[name="hour"]', '10');
    await page.check('input[name="gender"][value="M"]');
    await page.selectOption('select[name="category"]', 'daily');
    console.log('  ✓ 입력 완료');

    // 4. 제출 + 결과 대기
    console.log('[4/6] 운세보기 제출 (최대 120초 대기)...');
    // localStorage 캐시 클리어
    await page.evaluate(() => {
      for (let i = localStorage.length - 1; i >= 0; i--) {
        const k = localStorage.key(i);
        if (k && k.startsWith('fortunova_')) localStorage.removeItem(k);
      }
    });
    await page.click('button[type="submit"]');
    await page.waitForSelector('#result .fortune-reveal', { timeout: 120000 });
    console.log('  ✓ 결과 수신');

    // 5. 폼 접기 + 로딩 UI 확인
    console.log('[5/6] UX 검증...');
    const formCollapsed = await page.$eval('#form-section', el => el.classList.contains('collapsed'));
    console.log(`  ${formCollapsed ? '✓' : '✗'} 폼 접힘 상태: ${formCollapsed}`);

    const summaryVisible = await page.$eval('#form-summary', el => el.style.display !== 'none');
    console.log(`  ${summaryVisible ? '✓' : '✗'} 요약 바 표시: ${summaryVisible}`);

    // 6. 풍부한 결과 검증 (핵심!)
    console.log('[6/6] 풍부한 운세 결과 검증...');
    const resultHtml = await page.innerHTML('#result');
    const resultText = await page.$eval('#result', el => el.textContent);

    // 에러 체크
    if (resultHtml.includes('LLM_UNAVAILABLE') || resultHtml.includes('VALIDATION_ERROR')) {
      throw new Error(`에러 응답: ${resultHtml.substring(0, 300)}`);
    }
    if (resultHtml.includes('파싱에 실패')) {
      throw new Error('LLM 응답 파싱 실패');
    }

    const checks = [];
    function check(name, passed) {
      checks.push({ name, passed });
      console.log(`  ${passed ? '✓' : '✗'} ${name}`);
    }

    // 종합 점수
    const scoreMatch = resultText.match(/(\d+)\s*점/);
    check('종합 점수', !!scoreMatch);
    if (scoreMatch) console.log(`    → ${scoreMatch[1]}점`);

    // 요약 (summary)
    check('요약 (summary)', resultHtml.includes('font-serif') && resultHtml.includes('text-center'));

    // 상세 설명 (detail) - "오늘의 운세" 섹션
    check('상세 설명 (detail)', resultText.includes('오늘의 운세'));

    // 오행 흐름 (elementInsight)
    check('오행 흐름', resultText.includes('오행 흐름'));

    // 오늘의 조언 (advice)
    check('오늘의 조언', resultText.includes('오늘의 조언'));

    // 오늘의 팁 (dayTip)
    const hasDayTip = resultHtml.includes('✨');
    check('오늘의 팁', hasDayTip);

    // 세부운 그리드 (subFortunes)
    const hasSubFortunes = resultText.includes('세부 운세');
    check('세부 운세 그리드', hasSubFortunes);

    // 세부운 개별 확인
    check('재물운', resultText.includes('재물운'));
    check('건강운', resultText.includes('건강운'));
    check('연애운', resultText.includes('연애운'));
    check('직장운', resultText.includes('직장운'));

    // 오행 해설 (elementExplanation)
    check('오행 해설', resultText.includes('오행 해설'));

    // 행운 정보 (lucky)
    check('행운 정보', resultText.includes('행운'));
    check('행운의 색', resultText.includes('행운의 색'));
    check('행운의 숫자', resultText.includes('행운의 숫자'));

    // 주의사항 (cautions)
    check('주의사항', resultText.includes('주의사항'));

    // 3개월 운세 흐름 (monthlyTrend)
    check('3개월 운세 흐름', resultText.includes('3개월 운세 흐름'));

    // 궁합/인간관계 팁 (compatibilityTip)
    check('궁합 팁', resultText.includes('궁합') || resultText.includes('인간관계'));

    // 대운 해석 (majorFateInterpretation)
    check('대운 해석', resultText.includes('대운 해석'));

    // 격언 (proverb)
    check('격언', resultHtml.includes('proverb-card'));

    // 사주 정보
    check('사주 정보', resultText.includes('사주:'));

    // 남은 횟수
    check('남은 무료 횟수', resultText.includes('무료 횟수'));

    // 별점 (RatingStars)
    check('별점 표시', resultHtml.includes('★'));

    // 결과 요약
    const passed = checks.filter(c => c.passed).length;
    const total = checks.length;
    console.log(`\n📊 결과: ${passed}/${total} 항목 통과`);

    if (passed < total * 0.7) {
      // 스크린샷 + HTML 저장
      await page.screenshot({ path: '/tmp/fortunova-e2e-fail.png', fullPage: true });
      const fs = await import('fs');
      fs.writeFileSync('/tmp/fortunova-e2e-result.html', resultHtml);
      throw new Error(`풍부한 결과 검증 실패: ${passed}/${total} (70% 미달)`);
    }

    // 콘솔 에러
    if (errors.length > 0) {
      console.log(`\n⚠ 콘솔 에러 ${errors.length}건:`);
      errors.forEach(e => console.log(`  - ${e}`));
    }

    console.log('\n✅ E2E 검증 통과!');
  } catch (err) {
    console.error(`\n❌ E2E 검증 실패: ${err.message}`);
    await page.screenshot({ path: '/tmp/fortunova-e2e-fail.png', fullPage: true });
    console.error('  스크린샷: /tmp/fortunova-e2e-fail.png');
    process.exitCode = 1;
  } finally {
    await browser.close();
  }
}

verify();
